src/devices/bus/epson_qx/multifont.cpp

```diff
@@ -0,0 +1,284 @@
+{
+	I8039(config, m_mcu, 9_MHz_XTAL);
+	m_mcu->set_addrmap(AS_PROGRAM, &multifont_device::rom_map);
+	m_mcu->prog_out_cb().set(m_i8243, FUNC(i8243_device::prog_w));
+	m_mcu->p1_out_cb().set(FUNC(multifont_device::p1_w));
+	m_mcu->p2_out_cb().set(FUNC(multifont_device::p2_w));
+	m_mcu->bus_out_cb().set(FUNC(multifont_device::write_bus));
+	m_mcu->bus_in_cb().set(FUNC(multifont_device::read_bus));
+	m_mcu->t1_in_cb().set(FUNC(multifont_device::t1_r));
+	m_mcu->t0_in_cb().set(FUNC(multifont_device::t0_r));
+	I8243(config, m_i8243);
+	m_i8243->p4_out_cb().set(FUNC(multifont_device::rom_addr_w<0>));
+	m_i8243->p5_out_cb().set(FUNC(multifont_device::rom_addr_w<4>));
+	m_i8243->p6_out_cb().set(FUNC(multifont_device::rom_addr_w<8>));
+	m_i8243->p7_out_cb().set(FUNC(multifont_device::rom_select_w));
+}
+
+//-------------------------------------------------
+//  device_start - device-specific startup
+//-------------------------------------------------
+void multifont_device::device_start()
+{
+	m_iobase = 0x00;
+	m_data_out = 0x00;
+	m_data_in = 0x00;
+	m_p1 = 0xff;
+	m_p2 = 0xff;
+
+	save_item(NAME(m_iobase));
+	save_item(NAME(m_status));
+	save_item(NAME(m_data_out));
+	save_item(NAME(m_data_in));
+	save_item(NAME(m_p1));
+	save_item(NAME(m_p2));
+	save_item(NAME(m_rom_sel));
+	save_item(NAME(m_address));
+	save_item(NAME(m_rom_bank));
+	save_item(NAME(m_bus_reset));
+}
+
+//-------------------------------------------------
+//  device_reset - device-specific reset
+//-------------------------------------------------
+void multifont_device::device_reset()
+{
+	m_status = 0x00;
+	address_space &space = m_bus->iospace();
+
+	if (m_iobase != 0) {
+		space.unmap_readwrite(m_iobase, m_iobase+1);
+	}
+	m_iobase = (m_ioport->read() & 0x06) + 0xf8;
+
+	space.install_device(m_iobase, m_iobase+1, *this, &multifont_device::map);
+
+	m_bus_reset = true;
+}
+
+uint8_t multifont_device::t0_r()
+{
+	return m_status & 0x01;
+}
+
+uint8_t multifont_device::t1_r()
+{
+	return (m_status & 0x80) >> 7;
+}
+
+void multifont_device::write_bus(uint8_t data)
+{
+	if (!m_bus_reset) {
+		m_status = m_status | 0x80;
+		m_data_out = data;
+	} else {
+		m_bus_reset = false;
+	}
+}
+
+uint8_t multifont_device::read_bus()
+{
+	if (m_p1 & 0x01) {
+		m_status = (m_status & 0xfe);
+		return m_data_in;
+	} else if (m_p1 & 0x02) {
+		uint8_t *rom = m_fonts[m_rom_bank]->base();
+		return rom[m_address];
+	}
+	return 0xff;
+}
+
+void multifont_device::p1_w(uint8_t data)
+{
+	m_p1 = data;
+	if ((m_p1 & 0x80) == 0) {
+		get_slot()->intl_w(ASSERT_LINE);
+	}
+}
+
+void multifont_device::p2_w(uint8_t data)
+{
+	m_p2 = data;
+	m_i8243->p2_w(m_p2 & 0x0f);
+	update_bank_select();
+	m_status = (m_status & 0xfd) | ((m_p2 & 0x20) >> 4);
+}
+
+template<int Shift>
+void multifont_device::rom_addr_w(uint8_t data)
+{
+	m_address = (m_address & ~(0x00f << Shift)) | ((data & 0x0f) << Shift);
+}
+
+void multifont_device::rom_select_w(uint8_t data)
+{
+	m_rom_sel = (data & 0x0f) >> 2;
```

Comment:
This looks wrong like this – you’re selecting four least significant bits with the mask and then shifting right by two positions.  Two of the bits you selected with the mask are shifted off the end.

Reply:
This should be correct. I was masking the lower four bits since this is a 4 bit port. There should not actually be anything in the upper 4 bits, the & 0x0f was just to ensure they are empty. The only two bits i care about are bits 2 and 3.